compiler/resolution/lowerIterators.cpp

```diff
@@ -768,6 +768,10 @@ replaceIteratorFormalsWithIteratorFields(FnSymbol* iterator, Symbol* ic,
       // it is replaced by the field once the iterator class is created
       Expr* stmt = se->getStmtExpr();
 
+      if(toShadowVarSymbol(se->parentSymbol)) {
+        stmt = body->getFirstExpr();
+      }
+
```

Comment:
This one raised my eyebrow a bit from the perspective of "why haven't we been doing this already?".

Thinking out loud: In the past, shadow variables were only related to foralls where foralls are lowered into a complicated structure. In that scenario, we'd create shadow variables in the new AST we generate (in `coforall_fn`s, say) and that we wouldn't have to thread them through the actual iterator. But with shadow variables available on `ForLoop`s, that imply that the iterator itself needs to be handle shadow variables specifically.

Does that sound close to why we needed this here?

Reply:
Maybe...

I don't think I can answer your question but I'll explain the behavior I'm seeing to explain why I made this "fix":

On the line above we do: `Expr* stmt = se->getStmtExpr();`

And what I see is for a shadow variable is that `getStmtExpr` just returns `NULL`. Furthermore `se->parentExpr` returns NULL as well.

So I'm thinking there's either something wrong with how we're wiring these variables for `foreach` loops or (I think as you're suggesting) for `Forall` things get processed differently so we just don't encounter shadow variables for them in `replaceIteratorFormalsWithIteratorFields` at all.

My fix was a kludge: just special case this and explicitly set `stmt` to what seems like a safe statement.